[Antavo ] New Antavo Destination

bbantavo opened the pull request:
Added new Antavo Destination

## Testing

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)

joe-ayoub-segment commented:
Hi @bbantavo thanks for raising this PR for a new Integration. 

The code looks pretty straightforward. 

Instead of pointing out all the issues I'm going to ask you some broad questions, and based on your answers I might end up making some code changes myself before merging and deploying. 

1. Do you want to capture user events and user profile data for 'anonymous' users, as well as 'known' users? Or do you just want to capture data for 'known' users. A 'known' user is a user who has a userId which has been assigned to them by the customer. 
2. Do you want the event and profile Mappings to be automatically enabled when the Customer creates an instance of the Destination on the Segment platform (and provides auth credentials)?
3. Segment has recommended patterns for collecting user profile data. Customers tend to adhere to these patterns. Would you like me to update your code so that the 'Customer properties' field automatically maps to where the user profile data is likely to be?
4. Segment has a [specification for ecommerce events](https://segment.com/docs/connections/spec/ecommerce/v2/) which customers tend to adhere to. Does your platform have something similar? If so, it might be worth updating your code so that things like 'purchase amount', 'price', 'quantity' are automatically mapped?

Finally - would you please run the `yarn types` command, and commit the result? This should stop the 'Validate' CI check from failing. 

Thank you! 
Joe

joe-ayoub-segment commented on packages/destination-actions/src/destinations/antavo/index.ts:
```diff
@@ -0,0 +1,48 @@
+    },
+    testAuthentication: () => {
+      return true
+      // Return a request that tests/validates the user's credentials.
+      // If you do not have a way to validate the authentication fields safely,
+      // you can remove the `testAuthentication` function, though discouraged.
```
Hi @bbantavo is there a request you can make here to validate that the api_key is valid?

bbantavo replied:
hi @joe-ayoub-segment sadly we don't have a mechanism to validate the api_key as of now, I've removed `testAuthentication` from here

joe-ayoub-segment commented on packages/destination-actions/src/destinations/antavo/index.ts:
```diff
@@ -0,0 +1,48 @@
+    }
+  },
+
+  onDelete: async () => {
+    return true
+    // Return a request that performs a GDPR delete for the provided Segment userId or anonymousId
+    // provided in the payload. If your destination does not support GDPR deletion you should not
+    // implement this function and should remove it completely.
+  },
```
Hi @bbantavo 
Please remove the onDelete code if you don't intend to use it.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/antavo/index.ts:
```diff
@@ -0,0 +1,48 @@
+    fields: {
+      stack: {
+        label: 'Stack',
+        description: 'Antavo stack',
```
hi @bbantavo please indicate where the customer can get this value from.

bbantavo replied:
hi @joe-ayoub-segment updated the description to be more specific

joe-ayoub-segment commented on packages/destination-actions/src/destinations/antavo/index.ts:
```diff
@@ -0,0 +1,48 @@
+      },
+      api_key: {
+        label: 'API Key',
+        description: 'Antavo brand API key',
```
hi @bbantavo please indicate where the customer can get this value from.

bbantavo replied:
hi @joe-ayoub-segment updated the description to be more specific

joe-ayoub-segment commented on packages/destination-actions/src/destinations/antavo/profile/index.ts:
```diff
@@ -0,0 +1,99 @@
+  description: 'Sync profile updates into Antavo',
+  defaultSubscription: 'type = "identify"',
+  fields: {
+    customer: {
+      label: 'Customer ID',
+      description: 'User ID, selected in Antavo as customer identifier',
+      type: 'string',
+      required: true
+    },
```
Adding a default mapping so that userId is automatically mapped. 

```suggestion
    customer: {
      label: 'Customer ID',
      description: 'User ID, selected in Antavo as customer identifier',
      type: 'string',
      required: true,
      default: {
        '@path': '$.userId'
      }
    },
```

bbantavo replied:
hi @joe-ayoub-segment this is now added

joe-ayoub-segment commented on packages/destination-actions/src/destinations/antavo/profile/index.ts:
```diff
@@ -0,0 +1,99 @@
+          type: 'string'
+        },
+      },
+      default: {
+        first_name: '',
+        last_name: '',
+        email: '',
+        birth_date: '',
+        gender: '',
+        language: '',
+        phone: '',
+        mobile_phone: '',
+      }
```
Adding default mappings as per Segment identify spec https://segment.com/docs/connections/spec/identify/

```suggestion
       default: {
        first_name: {
          '@path': '$.traits.first_name'
        },
        last_name: {
          '@path': '$.traits.last_name'
        },
        email: {
          '@path': '$.traits.email'
        },
        birth_date: {
          '@path': '$.traits.birthday'
        },
        gender: {
          '@path': '$.traits.gender'
        },
        language: {
          '@path': '$.traits.language'
        },
        phone: {
          '@path': '$.traits.phone'
        },
        mobile_phone: {
          '@path': '$.traits.mobile_phone'
        }
      }
```

bbantavo replied:
hi @joe-ayoub-segment this is now added, thanks

joe-ayoub-segment commented on packages/destination-actions/src/destinations/antavo/event/index.ts:
```diff
@@ -0,0 +1,49 @@
+      label: 'Customer ID',
+      description: 'User ID, selected in Antavo as customer identifier',
+      type: 'string',
+      required: true
```
```suggestion
      required: true,
      default: {
        '@path': '$.userId'
      }
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/antavo/event/index.ts:
```diff
@@ -0,0 +1,49 @@
+      type: 'string',
+      required: false
+    },
+    data: {
+      label: 'Event data',
+      description: 'Event data',
+      type: 'object',
+      required: false
+    }
```
usually event data would be contained in this location in a Segment track() event. Up to you if you want to collect and send it. 

 default: {
        '@path': '$.properties'
 }

bbantavo replied:
hi @joe-ayoub-segment thanks for the suggestion! this field intentionally doesn't have a default mapping

joe-ayoub-segment commented:
hi @bbantavo I added some more comments - mostly default mappings for fields in your 2 Actions. 
Please also take a look at the testAuthencication() function and let me know if you can implement it or not. 
Finally, would you be able to write a couple of unit tests to validate that the payloads being sent to your platform are as expected please (for each Action)?

bbantavo commented:
> hi @bbantavo I added some more comments - mostly default mappings for fields in your 2 Actions. Please also take a look at the testAuthencication() function and let me know if you can implement it or not. Finally, would you be able to write a couple of unit tests to validate that the payloads being sent to your platform are as expected please (for each Action)?

hi @joe-ayoub-segment

I've added unit tests for `profile` and `event` action with a bunch of cases, please check!

joe-ayoub-segment commented:
hi @bbantavo thanks for updating the PR. This will go out with our next deploy which will probably be tomorrow (latest next Tuesday). 

Once deployed I'll share instructions for how to access and test the Integration. You might want to make some follow up changes after this, which is normal. 

Best regards,
Joe
